### PR TITLE
fix(drawer): fix drawer toggle race condition

### PR DESCRIPTION
**Pull request by ricmatsui**
Please provide enough information so that others can review your pull request:

**Motivation**

Explain the **motivation** for making this change. What existing problem does the pull request solve?

After navigating programmatically, the drawer is toggling unexpectedly if the JS thread is busy due to a race condition between the drawer animation and navigation state. Under extreme cases, the drawer may open and close multiple times before stopping.

This is because after the toggle animation is complete, the drawer always dispatches a navigation action. This is redundant in the case where the drawer opening or closing was done programmatically, and typically this should not cause another animation since the navigation state usually will not have changed. However, if the navigation state has already changed, the drawer may re-open (or close) and trigger another toggle animation.

This PR fixes this issue by only dispatching a navigation action when the drawer was toggled by the user and not programmatically.

**Test plan**

Demonstrate the code is solid. Example: the exact commands you ran and their output, screenshots / videos if the pull request changes UI.

Make sure you test on both platforms if your change affects both platforms.

The code must pass tests.

In this test, a slow component is added to cause the JS thread to be under heavy use during animation and the navigation state is logged:
```
git apply test.patch
```
<details>
  <summary>Contents of test.patch</summary>

```diff
diff --git a/example/src/index.tsx b/example/src/index.tsx
index 6be156ad..a1faf0d6 100644
--- a/example/src/index.tsx
+++ b/example/src/index.tsx
@@ -245,12 +245,13 @@ export default function App() {
       <NavigationContainer
         ref={navigationRef}
         initialState={initialState}
-        onStateChange={(state) =>
+        onStateChange={(state) => {
+          console.log(JSON.stringify(state.routes[0].state.history));
           AsyncStorage?.setItem(
             NAVIGATION_PERSISTENCE_KEY,
             JSON.stringify(state)
-          )
-        }
+          );
+        }}
         theme={theme}
         linking={{
           // To test deep linking on, run the following in the Terminal:
@@ -374,6 +375,14 @@ export default function App() {
                           }}
                         />
                         <Divider />
+                        {(() => {
+                          let j = 0;
+                          for (let i = 0; i < 1000000000; i++) {
+                            j += Math.random();
+                          }
+                          console.log('Slow component -- ' + j);
+                          return null;
+                        })()}
                         {(Object.keys(SCREENS) as (keyof typeof SCREENS)[]).map(
                           (name) => (
                             <List.Item

```

</details>

```
yarn example start
```

Before:

Opening and closing the drawer quickly causes the race condition and it opens and closes by itself repeatedly

https://user-images.githubusercontent.com/5288285/157373760-87b19ccf-5410-4eba-980d-dc312565193d.mov

After:

Opening and closing the drawer quickly does not cause any unexpected navigation

https://user-images.githubusercontent.com/5288285/157373855-362c47e5-6285-4243-9705-56ded862088b.mov

**Code formatting**

Look around. Match the style of the rest of the codebase. Run `yarn lint --fix` before committing.

### CODE REVIEW

**Comment by elmcapp** [comment]
I have similar issue. The issue for me is that when it pops back open.

File: packages/drawer/src/views/modern/Drawer.tsx [hunk 1]
```diff
@@ -339,7 +343,7 @@ export default function Drawer({
             {drawerType !== 'permanent' ? (
               <Overlay
                 progress={progress}
-                onPress={() => toggleDrawer(false)}
+                onPress={() => toggleDrawer(false, true)}
```
Comment (vonovak) [inline packages/drawer/src/views/modern/Drawer.tsx:346]: Hello and thanks for the PR!
Note that I'm not the person who will have a say in merging this PR, but I just wanted to point out that 

`toggleDrawer(false, true)`, or more genetally `someFunction(false, true)` reads pretty badly. I'd rather refactor that into an object with named keys.

Just my 2 cents 🙂

Comment (ricmatsui) [reply packages/drawer/src/views/modern/Drawer.tsx]: Thanks, agreed. I have made the change now.

**Comment by Rc85** [comment]
Having the same issue. If I open the drawer and go to a screen and once on that screen, I quickly open the drawer again to another screen. The drawer will reopen once it gets to the new screen.

**Comment by kastriotkastrati** [comment]
Having the same issue. It also makes any button that fires the events inconsistent. A close drawer button will think it's closing the drawer while the drawer is already closed for example.

**Comment by DarioQuintanilla2109** [comment]
Similar issue! https://github.com/react-navigation/react-navigation/issues/10608
DrawerNavigator V6 cant have StackNavigators as its screen component. The drawer doesnt close when switching tabs. Leads to very bad UI.

**Comment by satya164** [comment]
Hello, the PR got closed automatically due to some changes in the main branch. Would you be able to send a new PR? Thank you.

**Pull request closed without merging**